src/arch/mod.rs

```diff
@@ -9,3 +12,43 @@ pub mod aarch64;
 
 #[cfg(target_arch = "aarch64")]
 pub use self::aarch64::*;
+
+#[derive(Error, Debug)]
+#[error("Frequency detection failed")]
+pub struct FrequencyDetectionFailed;
+
+pub fn detect_freq_from_sysinfo() -> std::result::Result<u32, FrequencyDetectionFailed> {
+	debug!("Trying to detect CPU frequency using sysinfo");
+
+	let mut system = System::new();
+	system.refresh_cpu_frequency();
+
+	let frequency = system.cpus().first().unwrap().frequency();
+
+	if !system.cpus().iter().all(|cpu| cpu.frequency() == frequency) {
+		// Even if the CPU frequencies are not all equal, the
+		// frequency of the "first" CPU is treated as "authoritative".
+		eprintln!("CPU frequencies are not all equal");
+	}
+
+	// TODO: What can I do with the library's insistence of using u64 whereas I have to use u32?
```

Comment:
This is a general problem. I've filed an issue here: https://github.com/hermit-os/hermit-entry/issues/35

Reply:
Nevermind. it is a kHz value. So `try_into` is just fine.

Reply:
Resolved in 408d5ee98fbe8e4f7801280bb08a5edcdf0652e5 (will be pushed in a bit).